src/bms.lisp

```diff
@@ -0,0 +1,93 @@
+(defun bms-loop () {
+    (var vmin-limit)
+    (var vmax-limit)
+    (var tmax-limit)
+    (var tmin-limit -10)
+    (var cell-balance 0.5)
+    (var mosfet-tmax-limit)
+    (var config-update-time)
+    (var fault 0u32)
+    (var v-cell-support (eq (first (trap (get-bms-val 'bms-v-cell-min))) 'exit-ok))
+    (var bms-data-version-support (and (eq (first (trap (get-bms-val 'bms-data-version))) 'exit-ok) (> (get-bms-val 'bms-data-version) 0)))
+    (defun update-config () {
+        (setq vmin-limit (conf-get 'bms-vmin-limit-start))
+        (setq vmax-limit (conf-get 'bms-vmax-limit-start))
+        (setq tmax-limit (- (conf-get 'bms-t-limit-start) 3)) ; Start 3 degrees before Motor CFG -> BMS limiting functionality would happen.
+        (setq mosfet-tmax-limit (+ tmax-limit 15)) ; Set bms temp limit to +15C past cell limit
+        (setq config-update-time (systime))
+    })
+    (var fault-codes '(
+        (BMSF_NONE . 0)
+        (BMSF_CONNECTION . 1)
+        (BMSF_OVER_TEMP . 2)
+        (BMSF_CELL_OVER_VOLTAGE . 3)
+        (BMSF_CELL_UNDER_VOLTAGE . 4)
+        (BMSF_CELL_OVER_TEMP . 5)
+        (BMSF_CELL_UNDER_TEMP . 6)
+        (BMSF_CELL_BALANCE . 7)
+    ))
+    (defun set-fault (fault-code) {
+        (if (eq fault-code 'BMSF_NONE) {
+            (setq fault 0u32)
+        }{
+            (setq fault (bitwise-or fault (shl 1 (- (assoc fault-codes fault-code) 1))))
+        })
+    })
+    (update-config)
+    (loopwhile t {
+        (if (> (secs-since config-update-time) 1.0) {
+            (update-config)
+        })
+        (if (or (= vmin-limit 0.0) (= vmax-limit 0.0) (= tmax-limit 0.0)) {
```

Comment:
What is this condition for? It doesn't look right to skip all the checks if one of them is 0.

Reply:
Ok, I obviously didn't even read the last few messages when I came back with this review. But my message kind-of confirms this is confusing. The code is already running here, this check is not like an ordinary enable switch (in addition to it being hooked to three or-ed values). IMO a global in-package switch (as I described in the review) would be much better. Potentially any of the above being zero could be used to disable that particular check (for when e.g. the user has improperly configured thermistors or some other malfunction of one them).